src/gpu/amd/miopen_matmul.hpp

```diff
@@ -0,0 +1,176 @@
+/*******************************************************************************
+* Copyright 2020-2022 Intel Corporation
+* Copyright 2020 Codeplay Software Limited
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_AMD_MIOPEN_MATMUL_HPP
+#define GPU_AMD_MIOPEN_MATMUL_HPP
+
+#include <assert.h>
+
+#include "common/matmul_pd.hpp"
+#include "common/primitive.hpp"
+
+#include "gpu/amd/miopen_matmul_executor.hpp"
+#include "gpu/amd/miopen_matmul_impl.hpp"
+#include "gpu/amd/sycl_hip_utils.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace amd {
+
+struct miopen_matmul_t : public primitive_t {
+    using primitive_t::primitive_t;
+    struct pd_t : public matmul_pd_t {
+        using matmul_pd_t::matmul_pd_t;
+
+        DECLARE_COMMON_PD_T("hip:miopen:any", miopen_matmul_t);
+
+        status_t init(engine_t *) {
+            using namespace data_type;
+            using smask_t = primitive_attr_t::skip_mask_t;
+
+            data_type_t src_dt = src_md()->data_type;
+            data_type_t dst_dt = dst_md()->data_type;
+            data_type_t wei_dt = weights_md(0)->data_type;
+            data_type_t bia_dt
+                    = with_bias() ? weights_md(1)->data_type : data_type::f32;
+
+            bool f32_case = utils::everyone_is(f32, src_dt, wei_dt, dst_dt);
+            bool f16_case = utils::everyone_is(f16, src_dt, wei_dt, dst_dt);
+            bool s8_case
+                    = utils::everyone_is(s8, src_dt, wei_dt) && (dst_dt == s32);
+            bool bf16_case = utils::everyone_is(bf16, src_dt, wei_dt)
+                    && (dst_dt == f32);
```

Comment:
As per latest discussion keeping it as it is since we blocked bf16 bf16 bf16 cases since bias issue is there and below two tests are failing

./benchdnn --matmul --engine=gpu --skip-impl=ref --allow-enum-tags-only=false --dt=bf16:bf16:bf16 --wtag=bax --strides=::12x4x1 3x2x4:3x4x4:3x2x4
./benchdnn --matmul --engine=gpu --skip-impl=ref --allow-enum-tags-only=false --dt=bf16:bf16:bf16 --wtag=bax --strides=::21x7x1 3x2x4:3x4x4:3x2x4

Reply:
I tested the cases, turned out those are not data type specific. They fail for any data type. I figured the reason which is incorrect leading dimensions for A and B and batch strides for A and B. I modified the way we deduce those values and now all cases pass. So I enabled bf16bf16bf16 case.